src/mlpack/methods/ann/layer/multihead_attention.hpp

```diff
@@ -0,0 +1,422 @@
+/**
+ * @file methods/ann/layer/multihead_attention.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Definition of the MultiheadAttention class.
+ *
+ * @code
+ * @article{NIPS'17,
+ *   author  = {Ashish Vaswani, Llion Jones, Noam Shazeer, Niki Parmar,
+ *              Aidan N. Gomez, Jakob Uszkoreit, Łukasz Kaiser,
+ *              Illia Polosukhin},
+ *   title   = {Attention Is All You Need},
+ *   year    = {2017},
+ *   url     = {http://arxiv.org/abs/1706.03762v5}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_HPP
+#define MLPACK_METHODS_ANN_LAYER_MULTIHEAD_ATTENTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "layer_types.hpp"
+#include "softmax.hpp"
+#include "linear.hpp"
+#include "../init_rules/glorot_init.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Multihead Attention allows the model to jointly attend to information from
+ * different representation subspaces at different positions. With a single
+ * attention head, averaging inhibits this. [arxiv.org:1706.03762v5]
+ *
+ * @tparam AttentionPolicy The type of multihead attention to be used. Use
+ *         'SelfAttention' for self-attention and 'EncoderDecoderAttention' for
+ *         encoder-decoder attention.
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    class AttentionPolicy,
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class MultiheadAttention
+{
+ public:
+  //! Element Type of the input.
+  typedef typename InputDataType::elem_type ElemType;
+
+  /**
+   * Default constructor.
+   */
+  MultiheadAttention();
+
+  /**
+   * Create the MultiheadAttention object using the specified modules.
+   *
+   * @param tgtSeqLen The length of target sequence.
+   * @param srcSeqLen The length of the source sequence.
+   * @param embedDim Total dimension of the model.
+   * @param numHeads Number of parallel attention heads.
+   * @param dropout The dropout rate for attention output weights.
+   * @param deterministic If false, dropout layer is omitted else dropout layer
+   *        is applied with dropout rate `dropout`.
+   */
+  MultiheadAttention(const size_t tgtSeqLen,
+                     const size_t srcSeqLen,
+                     const size_t embedDim,
+                     const size_t numHeads,
+                     const ElemType dropout = 0.0,
```

Comment:
I guess it makes sense to use a value > 0 here, since dropout is enabled by default, but using a dropout value of 0 is the same as setting `deterministic = false`.

Reply:
Sure. I'm setting default value of dropout to `0.1`.